clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -275,8 +275,10 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
                 for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses().entrySet()) {
                     TopicPartition tp = entry.getKey();
                     ProduceResponse.PartitionResponse partResp = entry.getValue();
-                    ProducerBatch batch = batches.get(tp);
-                    completeBatch(batch, partResp, correlationId, now);
+                    if (tp != null && partResp != null) {
+                        ProducerBatch batch = batches.get(tp);
+                        completeBatch(batch, partResp, correlationId, now);
+                    }
```

Comment:
By the way, I have a suspicion this broker bug was fixed by https://github.com/apache/kafka/commit/1659ca1773596b5889fe8b2f163196f7b32ea70a#diff-a7cb05b8e918d46d237683e0f02756b9R179
There was a race condition where we would clear the map that was then used to populate the response if there was an unexpected broker exception.